src/mo_frontend/typing.ml

```diff
@@ -1884,12 +1896,24 @@ and check_exp' env0 t exp : T.typ =
     t
   | _ ->
     let t' = infer_exp env0 exp in
-    if not (T.sub t' t) then
+    let u, u' =
+      let open T in
+      if is_func t && is_func t' then
+        let transfer_arity from toh = match from, List.map normalize toh with
```

Comment:
maybe we shouldn't normalise here, as it exposes `Tup`, where none was before. We need the some tests, but also monomorphic ones to see if there are discrepancies.